grc: add support for absolute store paths

azahi opened the pull request:
## Description of changes

Add support for absolute paths from Nix store. This will allow grc to apply configs to `/nix/store/*-package-1.0.0/bin/hello` executables.
<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

azahi committed:
grc: add azahi as maintainer

Artturin commented on pkgs/tools/misc/grc/default.nix:
```diff
@@ -21,6 +21,10 @@ buildPythonApplication rec {
         --replace /usr/local/ $out/
     done
+
+    # Support for absolute store paths.
+    substituteInPlace grc.conf \
+      --replace "^([/\w\.]+\/)" "^([/\w\.\-]+\/)"
```
Needs a better comment/commit message explaining what is being done

azahi committed:
grc: add support for absolute store paths

Add support for absolute paths from Nix store. This will allow grc to
apply configs to direct execution of
`/nix/store/*-package-1.0.0/bin/hello` paths and not just PATH
expansion.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/2794

Artturin commented:
`grc $(chase $(whence ps)) aux`
and
`grc ps aux`

Have colors

```
$ whence ps
/etc/profiles/per-user/artturin/bin/ps
grc $(whence ps) aux
```

Doesn't but it's not a regression

Would be nice to tweak and upstream

Izorkin commented:
Is it possible to add a patch that changes the blue color to a brighter blue?
Example:
```diff
diff --git a/pkgs/tools/misc/grc/default.nix b/pkgs/tools/misc/grc/default.nix
index 19e926c5caf6..9fde688a3108 100644
--- a/pkgs/tools/misc/grc/default.nix
+++ b/pkgs/tools/misc/grc/default.nix
@@ -25,6 +25,11 @@ buildPythonApplication rec {
     # Support for absolute store paths.
     substituteInPlace grc.conf \
       --replace "^([/\w\.]+\/)" "^([/\w\.\-]+\/)"
+
+    # Replace blue color
+    substituteInPlace colourfiles/conf.* \
+      --replace "colours=blue" "colours=bright_blue" \
+      --replace "colours=default,blue" "colours=default,bright_blue"
   '';

   nativeBuildInputs = [ installShellFiles ];
```
Reason: оn a dark background it is very difficult to read text written in blue. You have to strain your eyesight.

peterhoeg commented:
┌────
│ {
│   environment.systemPackages = [
│     (grc.overrideAttrs (old: {
│       postPatch = old.postPatch + ''
│         for f in colourfiles/conf.* ; do
│           substituteInPlace $f --replace what-ever you-want-it-to-be
│         done
│       '';
│     }))
│   ];
│ }
└────

The better way is of course to patch upstream to look in `/etc/grc' (or similar) for additional configuration files, so you can override things there.

Izorkin commented:
Maybe then add an option - `replaceBlueColor ? false`.
Then it will be easier to override the configuration and anyone experiencing the same complexity can take advantage of this?

azahi commented:
@Izorkin This PR is completely unrelated to your request. Please create a separate issue or a PR if you want to see this added. 

It also seems like what you want is purely user-specific and depends entirely on the colourscheme, I'm not sure if we should patch something like this in Nixpkgs. I think the best solution here for you is to override the package locally just for yourself like in Peter's suggestion.
